Feature/question3-2/beautify #4

big-Miyabi opened the pull request:
課題
---
- [[課題3-2][UI]検索ボタンをリッチにする。Issue #4 · Caraquri/intern_2week_study_ios · GitHub](https://github.com/Caraquri/intern_2week_study_ios/issues/4)

概要
---
- 枠線やbuttonに色をつけて見やすいボタンにする

特に確認して頂きたい項目
---
- UIColorでの色の指定をRGBではなくHex値で指定できるようにコードを拡張した点
	- RGB値で指定する際、`red: x / 255`というように255で割らなければならず、可読性が低いと考えたため
	- Hex値をString型で定数宣言しておけば、定数名を見ればどのような色かおおよそ予想がつき、引数もRGB値を3つ渡さずにHex値一つで足りるので可読性が上がると考えたため
https://github.com/big-Miyabi/intern_2week_study_ios/blob/54cc9d590eee33da8c1d33c137f524d3c1ef427b/intern_2week_study/Utils/Extensions/UIColor%2BExtensions.swift#L3-L17
https://github.com/big-Miyabi/intern_2week_study_ios/blob/54cc9d590eee33da8c1d33c137f524d3c1ef427b/intern_2week_study/Question/Question3/Search/SearchViewController.swift#L31-L37<br/>
- 押下中の色を変更したかったため、`addTarget`メソッドを用いて、`.touchDown`時は色が暗くなり、`.touchUpInside`と`.touchUpOutside`時は色が明るくなる(元に戻る)というように処理を記述した
	- この記述法で問題ないか？また、メソッド名は適切か?
https://github.com/big-Miyabi/intern_2week_study_ios/blob/54cc9d590eee33da8c1d33c137f524d3c1ef427b/intern_2week_study/Question/Question3/Search/SearchViewController.swift#L6-L29
- textFieldの色をカスタムするために`TextField.swift`を作成した
	- ファイル名やクラス名は適切か？
- 見た目の処理を行うファイルとバックの処理を行うファイルは分割した方が良いか？


動作確認用 Screenshots
---
<img src="https://user-images.githubusercontent.com/47961006/90605433-4d881a80-e239-11ea-97ff-79602463f6e1.gif" width="300" />

参考URL
---
- [SwiftButtonの色を変える方法 - Qiita](https://qiita.com/goombeer/items/2ca4e109a091925f43f3)
- [swiftでUIViewに影（ドロップシャドウ）をつける | アプリ開発の覚書「魔人ツールズ」](https://majintools.com/2018/11/08/shadow/)
- [iOS StoryboardでUIViewの角丸を設定する : 雑食プログラミング備忘録](http://chicketen.blog.jp/archives/4907195.html)
- [Interface Builderでの指定色とコードでの指定色の違い - Qiita](https://qiita.com/imk2o/items/f0d029836dd7fff690c7)
- [【Swift4】UIColorを#ffffffなどの16進数の文字列（Hex）で指定する方法 | ニートに憧れるプログラム日記](https://program-life.com/493)
- [【Swift】TextFieldのスタイルを入力状態で変更する - Qiita](https://qiita.com/6bar10/items/ca9b2a05c8302505a99a)
- [Swift：UIColorを16進数カラーコードで初期化する - Qiita](https://qiita.com/Kyome/items/eae6216b13c651254f64)

big-Miyabi committed:
Add: 背景色とボタンの色、角丸を加えた

Change: 実機でテストしてみたかったのでbundle identifierなどを変更した

big-Miyabi committed:
Add: UIButtonに影を付与

big-Miyabi committed:
Add: ボタン押下中に色が変わるようにした

Change
storyboard上の色の設定をsRGBに統一し、swiftファイル内で色を変更した時とカラーコードに違いが出ないようにした

big-Miyabi committed:
Change:lightenButton関数をsearchButtonTapped関数に改名し、両者の内容を一つの関数内にまとめた

big-Miyabi committed:
Rename: darkenButtonをdarkenに、searchButtonTappedをlightenAndSearchに変更

big-Miyabi committed:
Merge branch 'feature/question3-1' into feature/question3-2/beautify
Fix: Conflictを解消するためにUIを修正した

big-Miyabi committed:
Merge branch 'feature/question3-1' into feature/question3-2/beautify

PRで指摘された内容を元に以下の変更をマージ
- Improve: tableView.registerの処理を@IBOutlet内のdidSetメソッドに移動
- Rename: ArticleListViewController.makeInstanceをmakeViewControllerに改名
- Improve: cellの高さをマジックナンバーではなく構造体で定数定義
- Improve: 'if let'を'guard let'に変更, 余分な改行を削除

big-Miyabi committed:
Fix: Conflictの解消

big-Miyabi committed:
Fix: ArticleListCell.xibをAuto Layoutsを使って調整

big-Miyabi committed:
Improve: titleを太文字、descriptionの文字は小さく色は灰色にした

big-Miyabi committed:
Improve: searchButtonの装飾部分の処理をbeautifySearchButtonメソッドにまとめた

big-Miyabi committed:
Remove: 余分そうな改行を削除

big-Miyabi committed:
Add: UIColor+Extensions.swiftを作成し、hexで色を指定できるようにした

big-Miyabi committed:
Improve: 前回コミットで作成した拡張コードを元に、色の指定を全てHex値で行うように書き換えた
Fix: UIColor+Extensions.swift内でコロンの位置や変数名に関するエラーが起きていたので修正

big-Miyabi committed:
Add: TextFieldの色を変更する処理を実装

big-Miyabi committed:
Add: 検索ボタン押下時にTextFieldの色が元に戻る処理を追加

msugai-cq commented on intern_2week_study/Question/Question3/Search/SearchViewController.swift:
```diff
@@ -3,20 +3,47 @@ import UIKit
     
     @IBOutlet weak var textField: UITextField!
-    @IBOutlet weak var searchButton: UIButton!
+    @IBOutlet weak var searchButton: UIButton! {
```
`UITextField`はカスタムクラスを作成し、実装しているのに対して`UIButton`はカスタムクラスを作っていない理由は何でしょうか?

big-Miyabi replied:
`UIButton`のカスタムクラスについては完全に粗漏でした。レビューありがとうございます！

修正を試みたのですが、slackの2weeks_study_forumチャンネルで質問を投げた通り、『**どうやってSearchButtonのカスタムクラスから、SearchViewControllerクラスのメソッドとSearchTextFieldのカスタムクラスのメソッドをそれぞれ呼び出すか**』という点で躓いております。

`addTarget`関連の処理を`SearchViewController`クラス内に全て記述してしまえば動かせますが、それでも良いのでしょうか？
`UIButton`のカスタムクラスを作るのであれば`addTarget`関連も全てその中に記述した方が良いと考え、現在格闘しているのですがかなり難航しています。

解決策があればお手数ですがご教授頂けると幸いです🙏💦
それ以外に指摘を受けた箇所に関しては全て修正しました！

- [質問: 異なるクラスのメソッドを呼び出すやり方について - HackM](https://hackmd.io/@Ryoooooga/B1w2bcoMw)
  - Slackで質問を投げた時のリンク
- [intern_2week_study_ios/SearchViewController.swift big-Miyabi/intern_2week_study_ios](https://github.com/big-Miyabi/intern_2week_study_ios/blob/f46a05de9757fde44c30e7f26df04d33a96ba6fc/intern_2week_study/Question/Question3/Search/SearchViewController.swift)
  - `addTarget`関連の処理を`SearchViewController`クラス内に全て記述した時のコード

msugai-cq reviewed (changes requested):
みました:eyes:
何点かコメントしているので確認お願いします:pray:

textField, searchButtonで違う実装をしているのが気になります。

> textFieldの色をカスタムするためにTextField.swiftを作成した
ファイル名やクラス名は適切か？

流石に名前が何を指しているか分からないほど汎用的なので、変えて欲しいです。`SearchTextField`でも

> 見た目の処理を行うファイルとバックの処理を行うファイルは分割した方が良いか？

MVPとか設計的な話でしたらもちろんそうした方が良いです。
しかし、今回の課題くらいでしたら分けなくて構いません。

big-Miyabi committed:
Rename: TextField.swift → SearchTextField.swift

big-Miyabi committed:
Improve: 色を定数に分けていたが、定数にせずに直接UIColor+Extensions内に記述した

big-Miyabi committed:
Rename: beautifyUI → setupButtonStyle
Remove: 余分なfinal修飾子を削除

big-Miyabi committed:
Add: SearchButton.swiftを作成してUIButtonをカスタムクラス化した
Rename: SearchButton.swift内のメソッドは元々beautifyUIという名前だったが、setupButtonStyleに名前変更した

big-Miyabi committed:
Improve: UITextFieldDelegateをextensionで分割した

big-Miyabi committed:
Improve: UIColor+Extensions.swiftをconvenience initを使ったやり方に書き換えた

big-Miyabi committed:
Move: addTarget関連の処理をSearchButtonクラス内に移動(まだ途中)

Todo
SearchButtonクラスのlightenAndSearchメソッド内で、SearchViewController内のsearchArgiclesメソッドとSearchTextField内のlightenメソッドを呼び出す処理を追加する1
